### PR TITLE
feature: implement the cdn manager

### PR DESCRIPTION
**Pull request by Starnop**
### Ⅱ. Does this pull request fix one issue?

fixes part of #366

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)

Added

### Ⅴ. Special notes for reviews

I would split this pull request into multiple commits for the convenience of review.

### CODE REVIEW

**Comment by Starnop** [comment]
```
#!/bin/bash -eo pipefail
gometalinter --disable-all --skip vendor --skip test -E deadcode ./...
supernode/daemon/mgr/cdn/path_util.go:28:1:warning: getMetaDataKeyFunc is unused (deadcode)
supernode/daemon/mgr/cdn/path_util.go:29:1:warning: getMd5DataKeyFunc is unused (deadcode)
supernode/daemon/mgr/cdn/path_util.go:30:1:warning: getUploadKeyFunc is unused (deadcode)
supernode/daemon/mgr/cdn/path_util.go:34:1:warning: getUploadRawFunc is unused (deadcode)
Exited with code 1
```

The CI not pass because there are some functions are unused, however, they will be used later. So I don't think it should block the process of review.

File: supernode/daemon/mgr/cdn/super_writer.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+					pieceNum:         curPieceNum,
+					pieceSize:        task.PieceSize,
+					pieceContentSize: int32(bb.Len()),
+					pieceContent:     bb,
+				}
+				logrus.Infof("send the protocolContent taskID: %s pieceNum: %d", task.ID, curPieceNum)
+
+				realFileLength += config.PieceWrapSize
+			}
+			close(jobCh)
```
Comment (lowzj) [inline supernode/daemon/mgr/cdn/super_writer.go]: `jobCh` should be closed before `wg.Wait`/return when reader is finished whether it's successful or failed.

**Comment by Starnop** [comment]
All mentioned above has been updated. PTAL.THX.

File: supernode/daemon/mgr/cdn/super_writer_util.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+	"github.com/dragonflyoss/Dragonfly/supernode/config"
+	"github.com/dragonflyoss/Dragonfly/supernode/store"
+
+	"github.com/sirupsen/logrus"
+)
+
+func calculateRoutineCount(httpFileLength int64, pieceSize int32) int {
+	routineSize := config.CDNWriterRoutineLimit
+	if httpFileLength <= 0 || pieceSize <= 0 {
```
Comment (lowzj) [inline supernode/daemon/mgr/cdn/super_writer_util.go]: When `httpFileLength` equals 0, it means that this is a empty file, it just need 1 goroutine to write into storage.

File: supernode/daemon/mgr/cdn/super_writer.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	var bb = &bytes.Buffer{}
+
+	// start writer pool
+	routineCount := calculateRoutineCount(httpFileLength, task.PieceSize)
+	var wg = &sync.WaitGroup{}
+	jobCh := make(chan *protocolContent)
+	cw.writerPool(ctx, wg, routineCount, jobCh)
+
+	for {
+		n, e := reader.Read(buf)
```
Comment (lowzj) [inline supernode/daemon/mgr/cdn/super_writer.go]: considering how to process an empty file.

**Comment by Starnop** [comment]
@lowzj /request a review again. 😄

File: supernode/daemon/mgr/cdn/super_writer_util.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	var resultBuf = &bytes.Buffer{}
+
+	// write piece header
+	var header = make([]byte, 4)
+	binary.BigEndian.PutUint32(header, getPieceHeader(pieceContSize, PieceSize))
+	resultBuf.Write(header)
+
+	// write piece content
+
+	pieceContent := make([]byte, bufferLength)
```
Comment (lowzj) [inline supernode/daemon/mgr/cdn/super_writer_util.go]: * remove useless space
* move this line into the body of if statement

Comment (Starnop) [reply supernode/daemon/mgr/cdn/super_writer_util.go]: Done.

**Pull request merged**